internal/batches/service/remote.go

```diff
@@ -96,6 +102,111 @@ func (svc *Service) CreateBatchSpecFromRaw(
 	return resp.CreateBatchSpecFromRaw.ID, nil
 }
 
+// UploadBatchSpecWorkspaceFile uploads workspace files to the server.
+func (svc *Service) UploadBatchSpecWorkspaceFile(workingDir string, batchSpecID string, steps []batches.Step) error {
+	for _, step := range steps {
+		for _, mount := range step.Mount {
+			if err := svc.handleWorkspaceFile(workingDir, mount.Path, batchSpecID); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (svc *Service) handleWorkspaceFile(workingDir, mountPath, batchSpecID string) error {
+	actualFilePath := filepath.Join(workingDir, mountPath)
+	info, err := os.Stat(actualFilePath)
+	if err != nil {
+		return err
+	}
+	if info.IsDir() {
+		dir, err := os.ReadDir(actualFilePath)
+		if err != nil {
+			return err
+		}
+		for _, dirEntry := range dir {
+			err := svc.handleWorkspaceFile(workingDir, filepath.Join(mountPath, dirEntry.Name()), batchSpecID)
+			if err != nil {
+				return err
+			}
+		}
+	} else {
+		if err = svc.uploadFile(workingDir, mountPath, batchSpecID); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (svc *Service) uploadFile(workingDir, mountPath, batchSpecID string) error {
+	// Create a pipe so the requests can be chunked to the server
+	pipeReader, pipeWriter := io.Pipe()
+	w := multipart.NewWriter(pipeWriter)
+
+	// Write in a separate goroutine to properly chunk the file content. Writing to the pipe lets us not have
+	// to put the whole file in memory.
+	go func() {
+		defer pipeWriter.Close()
+		defer w.Close()
+
+		if err := createFormFile(w, workingDir, mountPath); err != nil {
+			// TODO
```

Comment:
What shall we do?

Reply:
Handling errors in goroutines is one of the biggest gripes I have...

Reply:
Turns out this was easy with the `pipeWriter` - can just do `pipeWriter.CloseWithError`